pkg/arvo/lib/auto.hoon

```diff
@@ -0,0 +1,238 @@
+      [%tsbn *]  (change p.gen q.gen)
+      [%tscm *]
+    %+  replace
+      loop(gen p.gen)
+    |.(loop(gen q.gen, sut (~(busk ut sut) p.gen)))
+  ::
+      [%wtcl *]  (bell p.gen q.gen r.gen)
+      [%fits *]  (both p.gen wing+q.gen)
+      [%wthx *]  loop(gen wing+q.gen)
+      [%dbug *]  loop(gen q.gen)
+      [%zpcm *]  (both p.gen q.gen)
+      [%lost *]  loop(gen p.gen)
+      [%zpmc *]  (both p.gen q.gen)
+      [%zpts *]  loop(gen p.gen)
+      [%zpvt *]  !!
+      [%zpzp *]  ~
+      *
+    =+  doz=~(open ap gen)
+    ?:  =(doz gen)
+      ~_  (show [%c 'hoon'] [%q gen])
+      ~>  %mean.'play-open'
+      !!
+    loop(gen doz)
+  ==
+  ::
+  ++  replace
+    |=  [a=(unit [term type]) b=(trap (unit [term type]))]
+    ^-  (unit [term type])
+    ?~(a $:b a)
+  ::
+  ++  both
+    |=  [a=hoon b=hoon]
+    (replace loop(gen a) |.(loop(gen b)))
+  ::
+  ++  bell
+    |=  [a=hoon b=hoon c=hoon]
+    (replace loop(gen a) |.((replace loop(gen b) |.(loop(gen c)))))
+  ::
+  ++  spec-and-hoon
+    |=  [a=spec b=hoon]
+    (replace (find-type-in-spec sut a) |.(loop(gen b)))
+  ::
+  ++  change
+    |=  [a=hoon b=hoon]
+    (replace loop(gen a) |.(loop(gen b, sut (~(play ut sut) a))))
+  ::
+  ++  grow
+    |=  m=(map term tome)
+    =/  tomes  ~(tap by m)
+    |-  ^-  (unit [term type])
+    =*  outer-loop  $
+    ?~  tomes
+      ~
+    =/  arms  ~(tap by q.q.i.tomes)
+    |-  ^-  (unit [term type])
+    =*  inner-loop  $
+    ?~  arms
+      outer-loop(tomes t.tomes)
+    %+  replace
+      loop(gen q.i.arms, sut (~(play ut sut) gen))
+    |.  inner-loop(arms t.arms)
+  --
+::
+::  Not implemented yet.  I wonder whether we should modify types found
+::  in spec mode such that if it's a mold that produces a type, it
+::  should just display the type and not that it's technically a
+::  function.
+::
+++  find-type-in-spec
+  |=  [sut=type pec=spec]
+  ^-  (unit [term type])
+  !!
+::
+++  replace-hax
+  |=  code=tape
+  %+  scan  code
+  %+  cook
+    |=  res=(list $@(@ [~ (unit term)]))
+    %-  trip
+    %-  crip
+    %+  turn  res
+    |=  elem=$@(@ [~ (unit term)])
+    ?@  elem
+      elem
+    ?~  +.elem
+      'magic-spoon'
+    (cat 3 'magic-spoon.' u.+.elem)
+  %-  star
+  ;~  pose
+    ;~  less
+      hax
+      prn
+    ==
+  ::
+    (stag ~ ;~(pfix hax (punt sym)))
+  ==
+::
+++  autoadvance
+  |=  [sut=type gen=hoon]
+  %+  bind  (find-type-mule sut gen)
+  |=  [id=term typ=type]
+  (longest-match (search-prefix id (get-identifiers typ)))
+::
+++  auto-advance
+  |=  [sut=type code=cord]
+  (autoadvance sut (scan (replace-hax (trip code)) vest))
+::
+++  tablist
+  |=  [sut=type gen=hoon]
+  %+  bind  (find-type-mule sut gen)
+  |=  [id=term typ=type]
+  (search-prefix id (get-identifiers typ))
+::
+++  tab-list
+  |=  [sut=type code=cord]
+  %+  bind  (find-type sut (scan (replace-hax (trip code)) vest))
+  |=  [id=term typ=type]
+  (search-prefix id (get-identifiers typ))
```

Comment:
Could you add some comments detailing these different entrypoints and their basic strategies (ie, `#` vs `magic-spoon`).

Reply:
Refactored and added some comments, lmk if it makes sense.